File: udemy11imp/one_away.py
def is_one_away_same_length(s1, s2):
	count_diff = 0
	for i in range(len(s1)):
		if s1[i]!=s2[i]:
			count_diff += 1
			if count_diff > 1:
				return False
	return True

#Assumption s1.length == s2.length + 1
def is_one_away_diff_length(s1,s2):
	i = 0
	count_diff = 0
	while i < len(s2):
		if s1[i + count_diff] == s2[i]:
			i += 1
		else:
			count_diff += 1
			if count_diff > 1:
				return False
	return True

def is_one_away(s1, s2):
	if len(s1) - len(s2) >=2 or len(s2) - len(s1) >=2:
		return False
	elif len(s1) == len(s2):
		return is_one_away_same_length(s1, s2)
	elif len(s1) > len(s2):
		return is_one_away_diff_length(s1, s2)
	else:
		return is_one_away_diff_length(s1, s2)


# NOTE: The following input values will be used for testing your solution.
print(is_one_away("abcde", "abcd"))  # should return True
print(is_one_away("abde", "abcde"))  # should return True
print(is_one_away("a", "a"))  # should return True
print(is_one_away("abcdef", "abqdef"))  # should return True
print(is_one_away("abcdef", "abccef"))  # should return True
print(is_one_away("abcdef", "abcde"))  # should return True
print(is_one_away("aaa", "abc"))  # should return False
print(is_one_away("abcde", "abc"))  # should return False
print(is_one_away("abc", "abcde"))  # should return False
print(is_one_away("abc", "bcc"))  # should return False

File: udemy11imp/oneAway.java
public class oneAway{
	public static void main(String[] args){
		System.out.println(isOneAway("abcde", "abcd"));  // should return true
        System.out.println(isOneAway("abde", "abcde"));  // should return true
        System.out.println(isOneAway("a", "a"));  // should return true
        System.out.println(isOneAway("abcdef", "abqdef"));  // should return true
        System.out.println(isOneAway("abcdef", "abccef"));  // should return true
        System.out.println(isOneAway("abcdef", "abcde"));  // should return true
        System.out.println(isOneAway("aaa", "abc"));  // should return false
        System.out.println(isOneAway("abcde", "abc"));  // should return false
        System.out.println(isOneAway("abc", "abcde"));  // should return false
        System.out.println(isOneAway("abc", "bcc"));  // should return false
	}

	public static Boolean isOneAway(String s1, String s2){
		if (s1.length() - s2.length() >= 2 || s2.length() - s1.length() >= 2){
			return false;
		} else if (s1.length() == s2.length()){
			return isOneAwaySameLength(s1, s2);
		} else if (s1.length() > s2.length()){
			return isOneAwayDiffLength(s1, s2);
		} else {
			return isOneAwayDiffLength(s2, s1);
		}
	}

	public static Boolean isOneAwaySameLength(String s1, String s2){
		int countDiff = 0;
		for (int i = 0; i < s1.length(); i++){
			if (s1.charAt(i) != s2.charAt(i)){
				countDiff += 1;
				if (countDiff > 1) return false;
			}
		}
		return true;
	}

	public static Boolean isOneAwayDiffLength(String s1, String s2){
		int i = 0;
		int countDiff = 0;
		while (i < s2.length()){
			if (s1.charAt(i + countDiff) == s2.charAt(i)){
				i += 1;
			} else {
				countDiff += 1;
			}
			if (countDiff > 1){
				return false;
			}
		}
		return true;
	}
}
